### PR TITLE
factor out common workflows

### PR DESCRIPTION
**Pull request by lsf37**
With [reusable GitHub workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) we can now factor out common workflow files and maintain them only once here in the `ci-actions` repo.

Should also make it easier to deploy additional PR label actions to many/all of the seL4 org repos.

Unfortunately they *have* to live in the somewhat hidden `.github/workflows` directory, which I find a bit awkward, but it's still better than having to replicate everything. 

I have so far only tested the workflows that don't need secrets and trigger on `pull_request_target`, but those work. The display name changes, which mean we will have to update the names of required checks in all repos, but the benefit shoudl be worth the one-time effort.

### CODE REVIEW

**Comment by lsf37** [comment]
Have now also tested `pull_request_target`, and it works. The only thing that is missing is testing secrets which only really works if the caller workflow has been merged.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:37]: Why does bench habe "freeze" and this one not?

File: .github/workflows/pr.yml
Comment (axel-h) [inline .github/workflows/pr.yml:34]: Could this use local paths now?

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:40]: We should remove the support for "push", as this workflow is not intended to be call on this event.

File: .github/workflows/pr.yml
Comment (axel-h) [reply .github/workflows/pr.yml:34]: Seems this does not work as relative paths are resolved against the caller's repo. :(

File: .github/workflows/sel4bench-pr.yml
Comment (axel-h) [inline .github/workflows/sel4bench-pr.yml:1]: Let's call this 'sel4bench-hw.yml` then to align the naming. I expect more "hw" and "sim" here for the other projects. Maybe hw/sim could become a "mode" parameter to project workflows?

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:37]: Probably copied from different sources. They should both have the "freeze" step so that we can be sure that all sub-tests are working on the same revisions.

File: .github/workflows/pr.yml
Comment (lsf37) [reply .github/workflows/pr.yml:34]: yes, correct

File: .github/workflows/sel4bench-pr.yml
Comment (lsf37) [reply .github/workflows/sel4bench-pr.yml:1]: I don't think sim/hw should be a parameter, because they have very different requirements and are fairly different workflows. Also simulation workflows, if they are not too many, could run by default without label, and the hardware ones should always be guarded on pull requests.

Comment (lsf37) [reply .github/workflows/sel4bench-pr.yml:1]: For `sel4bench-hw` name: I'm not sure, actually. I also first changed it to `sel4bench-hw`, but then changed it back, because all `sel4bench` runs are on hardware, and the distinguishing factor for this workflow (as opposed to other sel4bench hardware workflows like the deployment one) is that it runs from pull requests.

Maybe the name consistency in the calling context is better than this reason, but `sel4bench-hw` doesn't really make much sense as a name in itself.

Comment (lsf37) [reply .github/workflows/sel4bench-pr.yml:1]: let's just do the consistency thing and see how it goes, we can always rename it back.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:49]:
```suggestion
    # To reduce the load (especially on the machine queue) we cancel any older
```

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:71]: We have to add `strategy.job-index` here again, otherwise the matrix runs kill their silibings.
```suggestion
      group: ${{ github.workflow }}-sel4bench-hw-run-pr-${{ github.event.number }}-${{ strategy.job-index }}
```

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:47]: same here:
```suggestion
      group: ${{ github.workflow }}-sel4bench-build-pr-${{ github.event.number }}}-${{ strategy.job-index }}
```

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:52]: same here:
```suggestion
      group: ${{ github.workflow }}-sel4test-build-pr-${{ github.event.number }}}-${{ strategy.job-index }}
```

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:101]: same here:
```suggestion
      group: ${{ github.workflow }}-sel4test-hw-run-pr-${{ github.event.number }}-${{ strategy.job-index }}
```

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [inline .github/workflows/sel4test-sim.yml:18]: Debugging gives me `seL4Test-sel4sim-pr-` here when a never job cancels an old run. Debugging says
```
  "event_name": "pull_request",
    "event": {
      "action": "synchronize",
      ...
      "number": 140,      
      ...
```
so I'm a but puzzled why the PR number is missing there.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:92]: Since we have filtered almost everything in the `code` job, checking this seems enough:
```suggestion
    if: ${{ github.repository_owner == 'seL4' &&
            contains(github.event.pull_request.labels.*.name, 'hw-test')  }}
```

This is the context for a "labeld" event. the new label is part of the array then already.
```
"github": {
  "event_name": "pull_request_target",
  "event": {
    "action": "labeled",
    "label": {
      ...
      "name": "hw-test",
      ...
    },
    "number": 140,
    "pull_request": {
      ...
      "labels": [
        {
          ...
          "name": "hw-test",
          ...          
        }
      ],
...
```

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [reply .github/workflows/sel4test-sim.yml:18]: Seems the whole `github.event` does not exist a this stage.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:31]: Isn't this redundant?

Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:31]: if the workflow is used correctly probably yes, but if we use a wrong trigger then the rest of the event context is different and not safe to access, so I though it better to make sure it does not trigger at all instead of triggering based on data that might or might not make sense.

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:92]: Are you sure the label wasn't in there before from being set previously? Because I remember doing exactly this debug print and had to specifically distinguish the two cases because in one of them it wasn't in the list.

Maybe they changed this in the meantime to simplify this case. It'd be nice to know for sure what we can rely on. Will have a look at what the docs say.

Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:92]: The [docs say](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=labeled#pull_request) that a labeled event must have a `pull_request` field which must have a `labels` field. I guess the only remaining question is if we are guaranteed that the new label is part of that `labels` field. If yes, we really can simplify.

Comment (axel-h) [reply .github/workflows/sel4test-hw.yml:92]: > Are you sure the label wasn't in there before from being set previously? 

Yes, I have trief this a couple of times as I was wondering exactly about that. Not sure if there is a guarantee, tough.

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [reply .github/workflows/sel4bench-hw.yml:71]: Actually, even this might not be enough if we are planning to have multiple repos calling this, where the workflow has the same name. We have to even add the `<org>/<repo>` here also to avoid the corner killing case where this is triggered by two PRs from different repos which happen to have the same ID.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [reply .github/workflows/sel4test-hw.yml:31]: I meat grouping the labeld/not-labeld closer together using this:
```
github.event_name == 'pull_request_target' &&
((github.event.action != 'labeled' &&
  (contains(github.event.pull_request.labels.*.name, 'hw-build') ||
   contains(github.event.pull_request.labels.*.name, 'hw-test'))) ||
 (github.event.action == 'labeled' &&
  (github.event.label.name == 'hw-build' ||
   github.event.label.name == 'hw-test')))

```

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:92]: I figured out why it is they way it is: if the event is `labeled` the user might be adding a label that is a different label, e.g. `docs`. In that case we don't want to re-trigger the test if it is already running (and the label `hw-test` would already be in the list of labels, because it's already been set before).

File: .github/workflows/sel4bench-hw.yml
Comment (lsf37) [reply .github/workflows/sel4bench-hw.yml:71]: Concurrency groups have [`repository` scope](https://github.com/orgs/community/discussions/78332), so we don't need to add that one at least. But we do need the job-index as you have it above.

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:31]: Yes, that would be fine. Maybe slightly easier to read.

File: .github/workflows/sel4test-sim.yml
Comment (lsf37) [reply .github/workflows/sel4test-sim.yml:18]: In which scenario is that happening? I.e. are you seeing this in the old job that is being cancelled or in the new job that is the reason for cancellation?

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [reply .github/workflows/sel4test-hw.yml:92]: True,  but this is properly filtered-out by the job `code` already - and I want to keep this part. So, when we arrive here in the job `hw-run`, it can only have happened due to a push or a new "hw_build"/"hw_test" label. So we can actually simply check if we have the "hw-test" set here.

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [reply .github/workflows/sel4bench-hw.yml]: Right, and I forgot we are executing in the caller's repo context here (and that's why relative paths here don't work either....)

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [reply .github/workflows/sel4test-sim.yml:18]: I'm seeing this in the old job, where github gives the cancellation reason. This is the only place where the group name is ever printed. It seems impossible to see the concurrency group for a running job anywhere. You can print it in a job/step, but the context there seems slightly different, because this runs on the runner. The workflow control and concurrency seems to run on an "action server", which does not see some dynamic parts of the contexts. The github documentation is quite thin here, but that is what I found in some github community postings.

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:92]: I don't think that's good. Why make assumptions if we can just test what we need to test? If we ever refactor or move this it'll work fine the way it is now, but not if we depend on context info from other jobs.

File: .github/workflows/sel4test-sim.yml
Comment (lsf37) [reply .github/workflows/sel4test-sim.yml:18]: Ok, the docs do say that `github`, `inputs`, and `vars` are available at the [workflow level](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency), and additionally `needs`, `strategy`, and `matrix` on the [job level](https://docs.github.com/en/actions/using-jobs/using-concurrency#using-concurrency-in-different-scenarios). So I think we are fine with this expression, it's just hard to debug as usual on GitHub.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [reply .github/workflows/sel4test-hw.yml:92]: Well, there are pros and cons. We depend on the build step anyway at the moment. In a bigger picture, it would be nicer if a caller would decide when to invoke which step, but I have not found nice was for this. So let's keep it as it is for now.

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [reply .github/workflows/sel4test-sim.yml:18]: Let's see if I get some answers here: https://github.com/orgs/community/discussions/107552

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:47]: one `}` too much
```suggestion
      group: ${{ github.workflow }}-sel4bench-build-pr-${{ github.event.number }}-${{ strategy.job-index }}
```

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:52]: one `}` too much
```suggestion
      group: ${{ github.workflow }}-sel4test-build-pr-${{ github.event.number }}}-${{ strategy.job-index }}
```

File: .github/workflows/sel4test-sim.yml
Comment (lsf37) [reply .github/workflows/sel4test-sim.yml:18]: I'm going to put it into the job level for now, so that we can merge this. Would still be good to figure out what exactly is available, though.

**Comment by lsf37** [comment]
I think this is ready to merge now. We can move the concurrency group for `sel4test-simp` back up to the workflow level when we get a definitive answer in <https://github.com/orgs/community/discussions/107552>.

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [reply .github/workflows/sel4test-sim.yml]: The other option could be adding another callable workflow layer here. The the job level concurrency group there will cover everything. It could also filter the events and invoke this workflow with a simple bool parameter for enabling hw-test.

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:24]: Apply the same as in `sel4test-hw.yml` here also:
```
    if: ${{ github.event_name == 'pull_request_target' &&
            (
              github.event.action != 'labeled' &&
              (contains(github.event.pull_request.labels.*.name, 'hw-build') ||
               contains(github.event.pull_request.labels.*.name, 'hw-test')) ||
              github.event.action == 'labeled' &&
              (github.event.label.name == 'hw-build' ||
               github.event.label.name == 'hw-test')
            ) }}
```

File: .github/workflows/sel4test-sim.yml
Comment (axel-h) [inline .github/workflows/sel4test-sim.yml:15]: It seem the "test with" in PRs works well for "seL4test", "seL4bench" and "sim". But I'm not fully understanding when we have to do a code freeze step an when not. Seems we have 3 different way how things are done for seL4test, seL4bench and sim. So a follow up task once this is merged should be aligning this better.

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:79]: We could make this a step in the `code` job also. It avoids spawning another runner VM.
```
jobs:
  code:
    ...
    outputs:
      ...
      matrix: ${{ steps.matrix.outputs.matrix }}
    steps:
    ...
    - id: matrix
      uses: seL4/ci-actions/sel4test-hw-matrix@master

    ...
  hw-run:
    ...
    needs: hw-build
```

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:71]:
```suggestion
        name: sel4test-images-${{ matrix.march }}-${{ matrix.compiler }}
```
I have done some experimenting with running both `sel4test` and `sel4bench`. Since they are all part of the same workflow now, there can be upload conflicts here. 
- `sel4bench` uploads as `images-${{ matrix.march }}`
- `sel4test` uploads as `images-${{ matrix.march }}-${{ matrix.compiler }}`

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:116]:
```suggestion
          name: images-sel4test-${{ matrix.march }}-${{ matrix.compiler }}
```

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:62]:
```suggestion
        name: images-sel4bench-${{ matrix.march }}
```

File: .github/workflows/sel4bench-hw.yml
Comment (axel-h) [inline .github/workflows/sel4bench-hw.yml:104]:
```suggestion
          name: images-sel4bench-${{ steps.plat.outputs.march }}
```

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [inline .github/workflows/sel4test-hw.yml:76]:
```suggestion
        name: kernel.elf-sel4test-${{ matrix.march }}-${{ matrix.compiler }}
```

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:71]: No, I don't think that's a good idea, those names are partially used in other scripts that people have and are used to. I don't see any benefit in changing them, they are local to their specific workflow and make sense in their context.

File: .github/workflows/sel4test-hw.yml
Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:79]: No. This is supposed to be factoring out existing workflows not a complete redesign of everything. Keeping this step separate keeps those concerns separated. The code job should get the code, not the matrix.

File: .github/workflows/sel4test-sim.yml
Comment (lsf37) [reply .github/workflows/sel4test-sim.yml:15]: `Test with` works with all of them. It's part of the checkout scripts they all call. Code freeze just eliminates a race condition where one test job might see a different manifest to another one (so we get the one manifest in the beginning and give it to all jobs).

File: .github/workflows/sel4test-hw.yml
Comment (axel-h) [reply .github/workflows/sel4test-hw.yml:71]: We have only one shared worflow `sel4test-hw.yml` in the sel4 kernel project now that triggers both "sel4test" and "sel4bench" . So all these artifacts end up together, see e.g. the artifact list. At the moment with just these it works because the naming pattern is different. If another project is added here, it need to use a better pattern. And if we consider running "sel4bench" for the compilers it will also clash badly. Luckly the upload action warns on duplicate names now instead of silently overwriting them.

Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:71]: The upload action actually fails now if there is a duplicate. It's true that they will end up in the same summary screen. We'd still need to track down what depends on this, so we should not just change artifact names if we don't need to. If it does become a problem in the future, we can still change things then (and it's then only one central workflow, so it's much easier to do).

We don't need to get everything perfect now, we won't achieve that anyway. We should observe this in action for a while and then see what we need.

Comment (lsf37) [reply .github/workflows/sel4test-hw.yml:71]: Ok, I can see how the UI will be confusing for the `images` part. I don't think those are used in automation outside this workflow (but kernel.elf might be). I'll try to double check on that first part and change those.

**Review by axel-h — APPROVED** [review]
It's good to merge this in the current state. I've done quite a lot of experimenting now and in general this works. So no gain in blocking the merge.

**Pull request merged**
